recaptcha_enterprise/demosite/src/main/java/app/MainController.java

```diff
@@ -36,68 +39,356 @@
+      }};
+      data = new HashMap<>() {{
+        put("data", result);
+      }};
+      return new ResponseEntity<>(data, httpHeaders, HttpStatus.OK);
+    } catch (Exception e) {
+      data = new HashMap<>() {{
+        put("data", new HashMap<>() {{
+          put("error_msg", e.toString());
+        }});
+      }};
+      return new ResponseEntity<>(data, httpHeaders, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  /**
+   * Return signup template.
+   */
+  @GetMapping(value = "/signup")
+  public static ModelAndView signup() {
+    return new ModelAndView("signup", CONTEXT);
   }
 
+  /**
+   * On signup button click, execute reCAPTCHA Enterprise assessment and take action according to
+   * the score.
+   */
+  @PostMapping(value = "/on_signup", produces = "application/json")
+  public static @ResponseBody ResponseEntity<HashMap<String, HashMap<String, String>>> onSignup(
+      @RequestBody Map<String, HashMap<String, String>> jsonData) {
+    final HttpHeaders httpHeaders = new HttpHeaders();
+    httpHeaders.setContentType(MediaType.APPLICATION_JSON);
+    String recaptchaAction = PROPERTIES.getProperty("recaptcha_action.signup");
+    HashMap<String, HashMap<String, String>> data;
+    Assessment assessmentResponse;
+    String verdict;
+
+    try {
+      // <!-- ATTENTION: reCAPTCHA Example (Server Part 1/2) Starts -->
+      assessmentResponse = CreateAssessment.createAssessment(
+          CONTEXT.get("project_id"),
+          CONTEXT.get("site_key"),
+          jsonData.get("recaptcha_cred").get("token"));
+
+      // Check if the token is valid, score is above threshold score and the action equals expected.
+      if (assessmentResponse.getTokenProperties().getValid() &&
+          assessmentResponse.getRiskAnalysis().getScore() > SAMPLE_THRESHOLD_SCORE &&
+          assessmentResponse.getTokenProperties().getAction().equals(recaptchaAction)) {
+        // Write new username and password to users database.
+        // String username = jsonData.get("recaptcha_cred").get("username");
+        // String password = jsonData.get("recaptcha_cred").get("password");
+        // Business logic.
+        // Classify the action as not bad.
+        verdict = "Not Bad";
+      } else {
+        // If any of the above condition fails, trigger email/ phone verification flow.
+        // Classify the action as bad.
+        verdict = "Bad";
+      }
+      // <!-- ATTENTION: reCAPTCHA Example (Server Part 1/2) Ends -->
+
+      // Return the risk score.
+      HashMap<String, String> result = new HashMap<>() {{
+        put("score", String.valueOf(assessmentResponse.getRiskAnalysis().getScore()));
+        put("verdict", verdict);
+      }};
+      data = new HashMap<>() {{
+        put("data", result);
+      }};
+      return new ResponseEntity<>(data, httpHeaders, HttpStatus.OK);
+    } catch (Exception e) {
+      data = new HashMap<>() {{
+        put("data", new HashMap<>() {{
+          put("error_msg", e.toString());
```

Comment:
Same question as above. Applies to all places we have this? Can we have them as seperate operations?
```
data = new HashMap<>();
dataMap = data.computeIfAbsent("data", d -> new HashMap<>())
dataMap.put("error_msg", e.toString());
```

Reply:
ditto.
I've updated it.